### PR TITLE
     add assert in immutable config provider constructor to make sure…

### PR DESCRIPTION
**Pull request by stevenzzzz**
… correct instance type gets set

Signed-off-by: Xin Zhuang <stevenzzz@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: add asserts in ImmutableConfigProviderBase too make sure correct instance type gets passed in.
Risk Level: LOW
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by stevenzzzz** [comment]
@AndresGuedez PTAL.

**Comment by stevenzzzz** [comment]
Added a test.

File: test/common/config/config_provider_impl_test.cc [hunk 1]
```diff
@@ -369,6 +369,31 @@ TEST_F(ConfigProviderImplTest, DuplicateConfigProto) {
...
+                            DummyConfigProviderManager& config_provider_manager,
+                            ConfigProviderInstanceType instance_type)
+      : ImmutableConfigProviderBase(factory_context, config_provider_manager, instance_type,
+                                    ApiType::Full) {}
+  ConfigConstSharedPtr getConfig() const override { return nullptr; }
+  std::string getConfigVersion() const override { return ""; }
+  const Protobuf::Message* getConfigProto() const override { return nullptr; }
+};
+
+TEST_F(ConfigProviderImplTest, AssertionFailureOnIncorrectInstanceType) {
```
Comment (AndresGuedez) [inline test/common/config/config_provider_impl_test.cc]: The test suite should be suffixed with `DeathTest` per https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#death-test-naming.

Comment (stevenzzzz) [reply test/common/config/config_provider_impl_test.cc]: PTAL.

File: test/common/config/config_provider_impl_test.cc [hunk 2]
```diff
@@ -369,6 +369,33 @@ TEST_F(ConfigProviderImplTest, DuplicateConfigProto) {
...
+class ConfigProviderImplDeathTest : public ConfigProviderImplTest {};
+
+TEST_F(ConfigProviderImplDeathTest, AssertionFailureOnIncorrectInstanceType) {
+  initialize();
+
+  InlineDummyConfigProvider foo(factory_context_, *provider_manager_,
+                                ConfigProviderInstanceType::Inline);
+  InlineDummyConfigProvider bar(factory_context_, *provider_manager_,
+                                ConfigProviderInstanceType::Static);
+  EXPECT_DEATH(InlineDummyConfigProvider(factory_context_, *provider_manager_,
```
Comment (AndresGuedez) [inline test/common/config/config_provider_impl_test.cc]: This needs to use EXPECT_DEBUG_DEATH().

Comment (stevenzzzz) [reply test/common/config/config_provider_impl_test.cc]: rigth

**Comment by stevenzzzz** [comment]
@htuch could you please give it a stamp.

**Pull request merged**